docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+* Bad, because it's a non-trivial departure from the current approach which will take more time to QA.
+
+This seems like a promising longer-term option, especially if we find that the majority of consumers
+are opting in to the fix proposed in this ADR. But the additional complexity weights heavily against
+trying to ship this as an initial fix for Project Nimbus.
+
+### Option D: Use a non-locking handlemap and rely on calling code to behave safely.
+
+Implement a variant of `ConcurrentHandleMap` that does not protect its members with a Mutex, which
+is the main source of constraints on concurrent execution in the current setup. Instead, this handlemap
+would hand out references without any runtime checks, on the assumption that the calling code is
+behaving in a safe manner. Replace all currently uses of `ConcurrentHandleMap` with this new non-locking variant.
+
+* Good, because this is a relatively small change from the current behaviour, which should be fairly
+  quick to ship.
+* Bad, because we lost Rust's compile-time safety guarantees.
+* Bad, because it's hard to communicate to consumers what "behave safely" really means (and we might not
+  even understand it ourselves).
+
+This seems to throw away some of the key safety benefits of using Rust, which makes it a very
+unattractive option.
+
+## Implementation Sketch
+
+### Add a `[Threadsafe]` attribute to UDL `interface` definitions.
+
+We would advise Nimbus SDK to update their UDL definition for `NimbusClient` like so:
+
+```idl
+[Threadsafe]
+interface NimbusClient {
+  // existing method definitions remain unchanged
+}
+```
+
+The [`uniffi_bindgen::interface::Object`](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L722) struct
+would grow a corresponding boolean `threadsafe` field and corresponding public accessor.
+When [building an instance of this struct from the UDL](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L786),
+we would inspect the list of attributes to look for one named "Threadsafe" and set the field
+if present. The way that we [handle the `[ByRef]` annotation on method arugments](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L658)
+will likely serve as a good example to follow.
+
+### Implement a less-locking HandleMap variant
+
+The `ffi_support` crate provides a basic non-locking [`HandleMap<T>`](https://docs.rs/ffi-support/0.4.0/ffi_support/handle_map/struct.HandleMap.html) struct,
+and it implements `ConcurrentHandleMap` as a thin wrapper around a `RwLock<HandleMap<Mutex<T>>>`. We can make our own variant that removes the inner mutex, as a thin wrapper around a`RwLock<HandleMap<T>>`. (The outer `RwLock` is still needed in order to guard mutations of the handlemap itself, and we must be careful to only hold it for short periods).
```

Comment:
Ref https://github.com/mozilla/uniffi-rs/pull/372#discussion_r558620397, this type is not sufficient because I don't think Rust will let us release the `RwLock` while there is an outstanding reference to an item from the map. As well it shouldn't: if we let one thread operate on an `&T` while a concurrent thread deletes the underlying `T` from the handlemap, we'd risk running `T::drop` while the `&T` was still alive.

I expect this will need to be `RwLock<HandleMap<Arc<T>>>` so that we can reference-count our way around this issue. Leaving a note for myself to update the final ADR before merge.